spring-integration-core/src/main/java/org/springframework/integration/router/AbstractMappingMessageRouter.java

```diff
@@ -189,15 +248,14 @@ public Collection<String> getDynamicChannelNames() {
 		return Collections.unmodifiableSet(this.dynamicChannels.keySet());
 	}
 
-	/**
-	 * Subclasses must implement this method to return the channel keys.
-	 * A "key" might be present in this router's "channelMappings", or it
-	 * could be the channel's name or even the Message Channel instance itself.
-	 * @param message The message.
-	 * @return The channel keys.
-	 */
-	protected abstract List<Object> getChannelKeys(Message<?> message);
-
+	@Override
+	protected void onInit() {
+		super.onInit();
+		Assert.state(!this.channelKeyFallback || !this.resolutionRequired || !this.defaultOutputChannelSet,
+				"The 'defaultOutputChannel' cannot be reached " +
+						"when both 'channelKeyFallback' & 'resolutionRequired' are set to true. " +
+						"See their javadocs for more information.");
+	}
```

Comment:
I don't think we can enforce this; what if both are true and channel resolution is successful.

Reply:
Correct. And what is the reason to have a `defaultOutputChannel` then?
You simply can set `channelKeyFallback = true`, but `resolutionRequired = false`.
So, if resolution is OK, we go there, if it is not we fallback to a default channel.

We probably can improve docs for this as well.